src/frontends/mutate.cpp

```diff
@@ -6,11 +6,96 @@ LICENSE file in the root directory of this source tree.
 */
 
 #include <loop_tool/mutate.h>
+#include <loop_tool/measure.hpp>
+#include <loop_tool/backend.h>
 
 #include <algorithm>
+#include <string>
 
 namespace loop_tool {
 
+std::vector<std::string> get_available_actions(const LoopTree& lt, LoopTree::TreeRef ref){
+  std::vector<std::string> avail_actions;
+
+  LoopTree::TreeRef p_ref = previous_ref(lt, ref);
+  LoopTree::TreeRef n_ref = next_ref(lt, ref);
+
+  if (p_ref == ref){
+    p_ref = -1;
+  }
+  if (n_ref == ref){
+    n_ref = -1;
+  }
+
+  // General operations
+  if (p_ref != -1){
+    avail_actions.push_back("up");
+  }
+  if (n_ref != -1){
+    avail_actions.push_back("down");
+  }
+
+  // Loop operations
+  if (lt.kind(ref) == LoopTree::LOOP){
+    // split
+    int max_vec_size = 1;
+
+    // the most inner-loop
+    if (all_of(lt.children(ref).begin(), lt.children(ref).end(), 
+          [&] (int i) {return lt.kind(i) == LoopTree::NODE;})
+        ){
+      max_vec_size = isa_traits<DABUN_ISA>().vector_size;
```

Comment:
So this is an interesting constraint that only loop_nest has. I wonder if we could soften it a bit and just fall back to default CPU when it's not met (super slow of course), as then the agent will end up "learning to vectorize" to some extent

Reply:
Yeah, I think the whole function get_available_actions should be implemented for each backend, and we should forward it here... I would like to discuss it on Zoom.